datafusion/common/src/dfschema.rs

```diff
@@ -412,17 +412,32 @@ impl DFSchema {
         }
     }
 
+    /// Check whether the column reference is ambiguous
+    pub fn check_ambiguous_name(
```

Comment:
I think the newly introduced functions `qualified_field_with_qualified_name` and `qualified_fields_with_qualified_name` do more than what we need. They might be unnecessary.
If I understand correctly, this function can be implemented as follows:
```rust
pub fn check_ambiguous_name(
        &self,
        qualifier: Option<&TableReference>,
        name: &str,
    ) -> Result<()> {
        let count = self
            .iter()
            .filter(|(field_q, f)| match (field_q, qualifier) {
                (Some(q1), Some(q2)) => q1.resolved_eq(q2) && f.name() == name,
                (None, None) => f.name() == name,
                _ => false,
            })
            .take(2)
            .count();
        if count > 1 {
            _schema_err!(SchemaError::AmbiguousReference {
                field: Column {
                    relation: None,
                    name: name.to_string(),
                },
            })
        } else {
            Ok(())
        }
    }
```

Reply:
Fixed. Thanks!

Reply:
It looks like we no longer need to add these two functions🤔, `qualified_field_with_qualified_name` and `qualified_fields_with_qualified_name`, along with the modifications to `qualified_field_with_name`

Reply:
These two functions will check the ambiguous column reference, it's necessary, or we could not identify the case below:

create table t1(v1 int);

select count(*)
from t1
right outer join t1
on t1.v1 > 0;

Reply:
We also need add `check_ambiguous_name` at [L132](https://github.com/apache/datafusion/blob/21e35eeb4e106835cc059205315d94ad0201a921/datafusion/sql/src/expr/identifier.rs#L146).

The changes to `qualified_field_with_name` will cause the following query to produce an incorrect result.
```sh
DataFusion CLI v41.0.0
> create table t1(v1 int) as values(100);
0 row(s) fetched.

> select t1.v1 from t1 join t1 using(v1) cross join (select struct('foo' as v1) as t1);
+--------+
| t1[v1] |
+--------+
| foo    |
+--------+
```
On the main branch, it gives
```sh
> select t1.v1 from t1 join t1 using(v1) cross join (select struct('foo' as v1) as t1);
+-----+
| v1  |
+-----+
| 100 |
+-----+
```
This is because it breaks [search_dfschema](https://github.com/apache/datafusion/blob/21e35eeb4e106835cc059205315d94ad0201a921/datafusion/sql/src/expr/identifier.rs#L274), making it unable to find `t1.v1`, but instead found the nested column `t1`.

Reply:
Is column reference 't1.v1' ambiguous in below query ?

`select t1.v1 from t1 join t1 using(v1) cross join (select struct('foo' as v1) as t1);`